src/main/java/dataMgmt/FixedSizeWritingQueue.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    /**
+     * Offer a patch for the queue to track and specify the updated absolute path
+     * of the file for if the patches need to be flushed to disc.
+     * @param patches the patches to offer to the queue
+     * @param absolutePath the new absolute path of the file tracked by this queue
+     * @return true if successfully added the patches, false if adding the patch failed
+     */
+    public boolean offerPatch(Patch[] patches, String absolutePath) {
+        if (!absolutePath.equals(this.absolutePath))
+            this.absolutePath = absolutePath;
+
+        boolean success = true;
+
+        for (Patch patch : patches) {
+            success = success && queue.offer(patch);
+        }
+
+        writePatchestoFileIfNeeded();
+
+        return success;
+    }
+
+    // applies the patches to the file if the queue length exceeds the WRITE_THRESHOLD
+    private void writePatchestoFileIfNeeded() {
+        if (queue.size() >= FileContentWriter.WRITE_THRESHOLD) {
+            synchronized (this) {
+                if (!writing) {
+                    List<Patch> patchesToWrite = new ArrayList<>();
+                    for (int i = 0; i < queue.size(); i++) {
+                        patchesToWrite.add(queue.get(i));
+                    }
+
+                    writing = true;
+                    new Thread(new FileWritingTask(absolutePath, patchesToWrite)).start();
+                }
+            }
+        }
+    }
+
+    /**
+     * A task for reading a file, applying a list of patches, and writing the file
+     */
+    protected class FileWritingTask implements Runnable {
+
+        // the file path of the file to patch
+        private String filePath;
+
+        // the list of patches to apply
+        private List<Patch> patches;
+
+        /**
+         * Creates a FileWritingTask that will apply the list of patches "patches" to the file at the location
+         * "filePath"
+         * @param filePath the path of the file to patch
+         * @param patches the patches to apply to the file
+         */
+        public FileWritingTask(String filePath, List<Patch> patches) {
+            this.filePath = filePath;
+            this.patches = patches;
+        }
+
+        @Override
+        public void run() {
+            String fileContents = "";
+            File file = new File(filePath);
+            try {
+                Scanner scanny = new Scanner(file).useDelimiter("\\Z");
+                if (scanny.hasNext())
+                    fileContents = scanny.next();
+            } catch (FileNotFoundException e) {
+                e.printStackTrace();
+                synchronized (this) {
+                    writing = false;
+                }
+                return;
+            }
+            String resultContents = patchManager.applyPatch(fileContents, patches);
+            try {
+                FileWriter writer = new FileWriter(file);
+                writer.write(resultContents);
```

Comment:
From what I can tell, this line threw a no pointer exception when running the tests locally but the test did not fail. I'm not sure if the exception is just because of gradle issues, but this failing should somehow be able to tell the test that it did not succeed